File: sqlens-frontend/lib/api/contracts/query-execution.ts
import { z } from "zod";

import { apiErrorSchema } from "@/lib/api/contracts/api-error";
import { queryInputSchema } from "@/lib/schemas/query";

/** Reuses the current sandbox's trusted query-input constraints at the API boundary. */
export const queryExecutionRequestSchema = queryInputSchema;

const queryValueSchema = z.union([
  z.string(),
  z.number().finite(),
  z.boolean(),
  z.null(),
]);

const queryStepRowSchema = z
  .object({
    id: z.string().min(1),
    data: z.record(queryValueSchema),
    status: z.enum(["keep", "filtered", "source", "joined"]),
    statusText: z.string().min(1),
    reason: z.string().min(1),
  })
  .strict();

const joinVisualizationDetailsSchema = z
  .object({
    leftTable: z.string().min(1),
    rightTable: z.string().min(1),
    leftKey: z.string().min(1),
    rightKey: z.string().min(1),
    joinType: z.string().min(1),
  })
  .strict();

/** Mirrors the existing SQLens visual-processing model at the transport boundary. */
const queryProcessingStepSchema = z
  .object({
    stepNumber: z.number().int().positive(),
    clause: z.string().min(1),
    title: z.string().min(1),
    badge: z.string().min(1),
    description: z.string().min(1),
    explanation: z.string().min(1),
    rowsBeforeCount: z.number().int().nonnegative(),
    rowsAfterCount: z.number().int().nonnegative(),
    rowEvaluations: z.array(queryStepRowSchema),
    joinDetails: joinVisualizationDetailsSchema.optional(),
  })
  .strict();

export const queryExecutionSuccessResponseSchema = z
  .object({
    status: z.literal("success"),
    query: z.string().min(1),
    executionTimeMs: z.number().finite().nonnegative(),
    columns: z.array(z.string().min(1)),
    rows: z.array(z.record(queryValueSchema)),
    totalRows: z.number().int().nonnegative(),
    steps: z.array(queryProcessingStepSchema),
  })
  .strict();

export const queryExecutionErrorResponseSchema = z
  .object({
    status: z.literal("error"),
    error: apiErrorSchema,
  })
  .strict();

export const queryExecutionResponseSchema = z.discriminatedUnion("status", [
  queryExecutionSuccessResponseSchema,
  queryExecutionErrorResponseSchema,
]);

export type QueryExecutionRequest = z.infer<typeof queryExecutionRequestSchema>;
export type QueryExecutionSuccessResponse = z.infer<
  typeof queryExecutionSuccessResponseSchema
>;
export type QueryExecutionErrorResponse = z.infer<
  typeof queryExecutionErrorResponseSchema
>;
export type QueryExecutionResponse = z.infer<
  typeof queryExecutionResponseSchema
>;

File: sqlens-backend/app/schemas/query_execution.py
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.common import ApiErrorResponse


class QueryExecutionRequest(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)

    query: str = Field(min_length=5, max_length=500)


class JoinVisualizationDetails(BaseModel):
    model_config = ConfigDict(extra="forbid")

    left_table: str = Field(min_length=1, serialization_alias="leftTable")
    right_table: str = Field(min_length=1, serialization_alias="rightTable")
    left_key: str = Field(min_length=1, serialization_alias="leftKey")
    right_key: str = Field(min_length=1, serialization_alias="rightKey")
    join_type: str = Field(min_length=1, serialization_alias="joinType")


class QueryStepRow(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str = Field(min_length=1)
    data: dict[str, str | int | float | bool | None]
    status: Literal["keep", "filtered", "source", "joined"]
    status_text: str = Field(min_length=1, serialization_alias="statusText")
    reason: str = Field(min_length=1)


class QueryProcessingStep(BaseModel):
    model_config = ConfigDict(extra="forbid")

    step_number: int = Field(gt=0, serialization_alias="stepNumber")
    clause: str = Field(min_length=1)
    title: str = Field(min_length=1)
    badge: str = Field(min_length=1)
    description: str = Field(min_length=1)
    explanation: str = Field(min_length=1)
    rows_before_count: int = Field(ge=0, serialization_alias="rowsBeforeCount")
    rows_after_count: int = Field(ge=0, serialization_alias="rowsAfterCount")
    row_evaluations: list[QueryStepRow] = Field(serialization_alias="rowEvaluations")
    join_details: JoinVisualizationDetails | None = Field(default=None, serialization_alias="joinDetails")


class QueryExecutionSuccessResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    status: Literal["success"]
    query: str = Field(min_length=1)
    execution_time_ms: float = Field(ge=0, serialization_alias="executionTimeMs")
    columns: list[str]
    rows: list[dict[str, str | int | float | bool | None]]
    total_rows: int = Field(ge=0, serialization_alias="totalRows")
    steps: list[QueryProcessingStep]


class QueryExecutionErrorResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    status: Literal["error"]
    error: ApiErrorResponse
